### ISSUE TITLE
How is KeepPokemonsThatCanEvolve used?

### ISSUE THREAD

**Issue by MaartenDeen** [NoxxDev/NecroBot#587]
I want Necrobot to keep pokemons that I can evolve. So when I have 125 Weedle candy, I want it to keep 10 Weedles.
I thought KeepPokemonsThatCanEvolve: true would do that, but it still only looks at KeepMinCp, KeepMinIvPercentage, KeepMinLvl
I had "KeepMinCp": 10, 6 Weedles and 122 candy. Then I put "KeepMinCp": 100 and it started transferring 3 Weedles that are below 100 CP:
[08:43:02] Weedle	- CP: 10  IV: 23.12%   [Best CP: 204  IV: 25.49%] (Candies: 123)
[08:43:25] Weedle	- CP: 73  IV: 15.57%   [Best CP: 204  IV: 25.49%] (Candies: 123)
[08:43:31] Weedle	- CP: 99  IV: 30.70%   [Best CP: 204  IV: 25.49%] (Candies: 125)
Now I have 3 Weedles and 125 candy. But I could have evolved all those three that were transferred.
What am I doing wrong?

My config.

  "PokemonConfig": {
    "CatchPokemon": true,
    "DelayBetweenPokemonCatch": 2000,
    "UseCatchLimit": true,
    "CatchPokemonLimit": 998,
    "CatchPokemonLimitMinutes": 1470,
    "UseIncenseConstantly": true,
    "UseEggIncubators": true,
    "UseEggIncubatorMinKm": 2,
    "UseLuckyEggConstantly": false,
    "UseLuckyEggsMinPokemonAmount": 30,
    "UseLuckyEggsWhileEvolving": false,
    "UseBerriesMinCp": 1000,
    "UseBerriesMinIv": 90.0,
    "UseBerriesBelowCatchProbability": 0.2,
    "UseBerriesOperator": "or",
    "MaxBerriesToUsePerPokemon": 1,
    "TransferWeakPokemon": true,
    "TransferDuplicatePokemon": true,
    "TransferDuplicatePokemonOnCapture": true,
    "RenamePokemon": false,
    "RenameOnlyAboveIv": false,
    "RenameTemplate": "{1}_{0}",
    "FavoriteMinIvPercentage": 95.0,
    "AutoFavoritePokemon": false,
    "MaxPokeballsPerPokemon": 6,
    "UseGreatBallAboveCp": 1000,
    "UseUltraBallAboveCp": 1250,
    "UseMasterBallAboveCp": 1500,
    "UseGreatBallAboveIv": 85.0,
    "UseUltraBallAboveIv": 95.0,
    "UseGreatBallBelowCatchProbability": 0.2,
    "UseUltraBallBelowCatchProbability": 0.1,
    "UseMasterBallBelowCatchProbability": 0.05,
    "AutomaticallyLevelUpPokemon": false,
    "OnlyUpgradeFavorites": true,
    "UseLevelUpList": true,
    "AmountOfTimesToUpgradeLoop": 5,
    "GetMinStarDustForLevelUp": 5000,
    "LevelUpByCPorIv": "iv",
    "UpgradePokemonCpMinimum": 1000.0,
    "UpgradePokemonIvMinimum": 95.0,
    "UpgradePokemonMinimumStatsOperator": "and",
    "EvolveAboveIvValue": 95.0,
    "EvolveAllPokemonAboveIv": false,
    "EvolveAllPokemonWithEnoughCandy": false,
    "EvolveKeptPokemonsAtStorageUsagePercentage": 110.0,
    "KeepPokemonsThatCanEvolve": true,
    "KeepMinCp": 100,
    "KeepMinIvPercentage": 50.0,
    "KeepMinLvl": 10,
    "KeepMinOperator": "or",
    "UseKeepMinLvl": false,
    "PrioritizeIvOverCp": false,
    "KeepMinDuplicatePokemon": 15,
    "UsePokemonToNotCatchFilter": false,
    "UsePokemonSniperFilterOnly": false,
    "DumpPokemonStats": false
  },

**Comment by Cranki** [comment]
"TransferWeakPokemon": true, << I think this command has higher prio than >> "KeepPokemonsThatCanEvolve: true", the weak pokemon command must be false, then bot keeps weak pokemon that can evolve.

**Comment by MaartenDeen** [comment]
I tried that, but that does not solve it completely. I now have 4 Squirtles, 24/25 candy:
cp 478 lvl 19 IV 43.7%
cp 191 lvl 8 IV 29.0%
cp 146 lvl 6 IV 24.1%
cp 62 lvl 3 IV 24.1%
And "KeepMinCp": 100, "KeepMinIvPercentage": 50.0, "KeepMinLvl": 10.
Why is the CP 62 Squirtle not transferred?

Config is below, no Squirtle in the PokemonsTransferFilter.

  "PokemonConfig": {
    "CatchPokemon": true,
    "DelayBetweenPokemonCatch": 2000,
    "UseCatchLimit": true,
    "CatchPokemonLimit": 998,
    "CatchPokemonLimitMinutes": 1470,
    "UseIncenseConstantly": true,
    "UseEggIncubators": true,
    "UseEggIncubatorMinKm": 2,
    "UseLuckyEggConstantly": false,
    "UseLuckyEggsMinPokemonAmount": 30,
    "UseLuckyEggsWhileEvolving": false,
    "UseBerriesMinCp": 1000,
    "UseBerriesMinIv": 90.0,
    "UseBerriesBelowCatchProbability": 0.2,
    "UseBerriesOperator": "or",
    "MaxBerriesToUsePerPokemon": 1,
    "TransferWeakPokemon": false,
    "TransferDuplicatePokemon": true,
    "TransferDuplicatePokemonOnCapture": false,
    "RenamePokemon": false,
    "RenameOnlyAboveIv": false,
    "RenameTemplate": "{1}_{0}",
    "FavoriteMinIvPercentage": 95.0,
    "AutoFavoritePokemon": false,
    "MaxPokeballsPerPokemon": 6,
    "UseGreatBallAboveCp": 1000,
    "UseUltraBallAboveCp": 1250,
    "UseMasterBallAboveCp": 1500,
    "UseGreatBallAboveIv": 85.0,
    "UseUltraBallAboveIv": 95.0,
    "UseGreatBallBelowCatchProbability": 0.2,
    "UseUltraBallBelowCatchProbability": 0.1,
    "UseMasterBallBelowCatchProbability": 0.05,
    "AutomaticallyLevelUpPokemon": false,
    "OnlyUpgradeFavorites": true,
    "UseLevelUpList": true,
    "AmountOfTimesToUpgradeLoop": 5,
    "GetMinStarDustForLevelUp": 5000,
    "LevelUpByCPorIv": "iv",
    "UpgradePokemonCpMinimum": 1000.0,
    "UpgradePokemonIvMinimum": 95.0,
    "UpgradePokemonMinimumStatsOperator": "and",
    "EvolveAboveIvValue": 95.0,
    "EvolveAllPokemonAboveIv": false,
    "EvolveAllPokemonWithEnoughCandy": false,
    "EvolveKeptPokemonsAtStorageUsagePercentage": 110.0,
    "KeepPokemonsThatCanEvolve": true,
    "KeepMinCp": 100,
    "KeepMinIvPercentage": 50.0,
    "KeepMinLvl": 10,
    "KeepMinOperator": "or",
    "UseKeepMinLvl": false,
    "PrioritizeIvOverCp": false,
    "KeepMinDuplicatePokemon": 15,
    "UsePokemonToNotCatchFilter": false,
    "UsePokemonSniperFilterOnly": false,
    "DumpPokemonStats": false
  },

**Comment by iwean** [comment]
I guess cp 62 squirtel is a new catch, so you must wait the bot back to the first sequence to transfer it. 

Or set to true >>> "TransferDuplicatePokemonOnCapture": false,

**Comment by MaartenDeen** [comment]
No, I waited and waited but it did not get transferred. I do not really TransferDuplicatePokemonOnCapture either. I've seen it transfer pokemons that IMHO also should not have been transferred.

**Comment by NoxxDev** [comment]
https://github.com/NoxxDev/NecroBot/wiki/Config#transferduplicatepokemon-value

`"KeepMinDuplicatePokemon": 15,`

**Comment by MaartenDeen** [comment]
Ok, hadn't seen that one.
But now it is not keeping to KeepPokemonsThatCanEvolve at all.
I had 5 Weedles and 52/12 candy. Good for evolving 4 Weedles. Necrobot chose to transfer 3 Weedles so I'm left with 2.

Maybe I don't understand the setting, maybe it's not what I'm looking for, but how can I prevent the bot from transferring pokemon that I can evolve?

**Comment by kalamera** [comment]
the "KeepPokemonsThatCanEvolve" does really work strange in 0.9.7 as it seems.

I have the "TransferWeakPokemon" set True which works well until I set the "KeepPokemonsThatCanEvolve" also True.
Then the bots keeps all the crappy pokemons, does not transfer any.

I would expect the bot to keep all the pokemons that fit the value "EvolveAboveIVValue" with "EvolveAllPokemonAboveIv" True.
When it collects "UseLuckyEggsMinPokemonAmount" of "Evolvable" pokomens It uses the Lucky Egg and starts evolving. hopefully not interrupted by the "Niatic servers unstable..." error message.

That worked well in 0.9.6 if I do recall correctly.
